### PR TITLE
[SPARK-16930][yarn] Fix a couple of races in cluster app initialization.

### PR DESCRIPTION
**Pull request by vanzin**
There are two narrow races that could cause the ApplicationMaster to miss
when the user application instantiates the SparkContext, which could cause
app failures when nothing was wrong with the app. It was also possible for
a failing application to get stuck in the loop that waits for the context
for a long time, instead of failing quickly.

The change uses a promise to track the SparkContext instance, which gets
rid of the races and allows for some simplification of the code.

Tested with existing unit tests, and a new one being added to test the
timeout code.

### CODE REVIEW

**Comment by vanzin** [comment]
@tgravescs

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala [hunk 1]
```diff
@@ -404,7 +410,8 @@ private[spark] class ApplicationMaster(
       clientMode = true)
     val driverRef = waitForSparkDriver()
     addAmIpFilter()
-    registerAM(rpcEnv, driverRef, sparkConf.get("spark.driver.appUIAddress", ""), securityMgr)
+    registerAM(sparkConf, rpcEnv, driverRef, sparkConf.get("spark.driver.appUIAddress", ""),
```
Comment (jerryshao) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: Maybe we could use `ConfigEntry` to replace here.

Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala]: Maybe, but that's an unrelated change.

**Comment by vanzin** [comment]
Ping, I'll push this tomorrow unless someone complains.

**Comment by vanzin** [comment]
Merging to master.
